tikv/batch_getter_test.go

```diff
@@ -0,0 +1,104 @@
+package tikv
+
+import (
+	"context"
+
+	. "github.com/pingcap/check"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/unionstore"
+)
+
+var _ = Suite(&testBatchGetterSuite{})
+
+type testBatchGetterSuite struct{}
+
+func (s *testBatchGetterSuite) TestBufferBatchGetter(c *C) {
+	snap := newMockStore()
+	ka := []byte("a")
+	kb := []byte("b")
+	kc := []byte("c")
+	kd := []byte("d")
+	snap.Set(ka, ka)
+	snap.Set(kb, kb)
+	snap.Set(kc, kc)
+	snap.Set(kd, kd)
+
+	buffer := unionstore.NewUnionStore(snap)
+	buffer.GetMemBuffer().Set(ka, []byte("a1"))
+	buffer.GetMemBuffer().Delete(kb)
+
+	batchGetter := NewBufferBatchGetter(buffer.GetMemBuffer(), snap)
+	result, err := batchGetter.BatchGet(context.Background(), [][]byte{ka, kb, kc, kd})
+	c.Assert(err, IsNil)
+	c.Assert(len(result), Equals, 3)
+	c.Assert(string(result[string(ka)]), Equals, "a1")
+	c.Assert(string(result[string(kc)]), Equals, "c")
+	c.Assert(string(result[string(kd)]), Equals, "d")
+}
+
+type mockBatchGetterStore struct {
+	index [][]byte
+	value [][]byte
+}
+
+func newMockStore() *mockBatchGetterStore {
+	return &mockBatchGetterStore{
+		index: make([][]byte, 0),
+		value: make([][]byte, 0),
+	}
+}
+
+func (s *mockBatchGetterStore) Len() int {
+	return len(s.index)
+}
+
+func (s *mockBatchGetterStore) Get(ctx context.Context, k []byte) ([]byte, error) {
+	for i, key := range s.index {
+		if kv.CmpKey(key, k) == 0 {
+			return s.value[i], nil
+		}
+	}
+	return nil, tikverr.ErrNotExist
+}
+
+func (s *mockBatchGetterStore) BatchGet(ctx context.Context, keys [][]byte) (map[string][]byte, error) {
+	m := make(map[string][]byte)
+	for _, k := range keys {
+		v, err := s.Get(ctx, k)
+		if err == nil {
+			m[string(k)] = v
+			continue
+		}
+		if tikverr.IsErrNotFound(err) {
+			continue
+		}
+		return m, err
+	}
+	return m, nil
+}
+
+func (s *mockBatchGetterStore) Set(k []byte, v []byte) error {
+	for i, key := range s.index {
+		if kv.CmpKey(key, k) == 0 {
+			s.value[i] = v
+			return nil
+		}
+	}
+	s.index = append(s.index, k)
+	s.value = append(s.value, v)
+	return nil
+}
+
+func (s *mockBatchGetterStore) Delete(k []byte) error {
+	s.Set(k, []byte{})
+	return nil
+}
+
+func (s *mockBatchGetterStore) Iter(k []byte, upperBound []byte) (unionstore.Iterator, error) {
```

Comment:
Could we remove this function?

Reply:
This method is implemented because the parameters of the `NewUnionStore` method need to implement the `uSnapshot` interface:
```Go
// uSnapshot defines the interface for the snapshot fetched from KV store.
type uSnapshot interface {
	// Get gets the value for key k from kv store.
	// If corresponding kv pair does not exist, it returns nil and ErrNotExist.
	Get(ctx context.Context, k []byte) ([]byte, error)
	// Iter creates an Iterator positioned on the first entry that k <= entry's key.
	// If such entry is not found, it returns an invalid Iterator with no error.
	// It yields only keys that < upperBound. If upperBound is nil, it means the upperBound is unbounded.
	// The Iterator must be Closed after use.
	Iter(k []byte, upperBound []byte) (Iterator, error)

	// IterReverse creates a reversed Iterator positioned on the first entry which key is less than k.
	// The returned iterator will iterate from greater key to smaller key.
	// If k is nil, the returned iterator will be positioned at the last key.
	// TODO: Add lower bound limit
	IterReverse(k []byte) (Iterator, error)
}
```
In fact, I just need to create a `MemDB` object. Can we consider exposing the `newMemDB` method?

Reply:
How about using `mockBatchGetterStore` as the first argument for the function `NewBufferBatchGetter` since the `mockBatchGetterStore` already implement the interface `BatchBufferGetter`, and also this would make the test easier to understand(the reviewer needn't understand what is `UnionStore` which could not be related to this test).

Reply:
This is a great idea @AndreMouche